### DISCUSSION TITLE
Poe script task uses global python path instead of the one in the virtual environment

### DISCUSSION THREAD

**Discussion by fritz-trawa** [nat-n/poethepoet#232] [Q&A] [answered]
```bash
pyenv global 3.12
python -V 
# 3.12
cd my-project
pyenv local 3.10
poetry install --no-root
python -V 
# 3.10
```

Then I have a simple python script to do the same

```python
# version.py
import sys

def echo():
    print("Python version:", sys.version)

if __name__ == "__main__":
    echo()
```

Then in my tasks file I have

```yaml
[tool.poe.tasks.version]
script = "version:echo"
```

Then when I test it

```bash
python version.py
# Python version: 3.10.14 ....

python -m version
# Python version: 3.10.14 ....

poe version
# Poe => version
# Python version: 3.12.4 ...
```

How do I get poe to use the version in my local virtual environment? Thanks

**Comment by nat-n** [comment]
Hi @fritz-trawa, thanks for the question.

I need some more info to tell whether what you describe is expected behavior or not. If used with poetry then poe will by default use whatever python is referenced in the virtualenv managed by poetry, that is the env returned by `poetry env info -p`. Running `poetry run python version.py` should confirm whether poetry is using the expected python version for its virtualenv; could you try that please?

Another important question, is the version task defined in the main pyproject.toml file alongside poetry related config or is it imported from another file? If it's an import, does it start working if you move it to the pyproject.toml? If so then we have a bug.

This is because poe will only default to using the [PoetryExecutor](https://github.com/nat-n/poethepoet/blob/32689ca6252d0da09232087ad8f97ea9337036eb/poethepoet/executor/poetry.py) if it detects poetry related config, which should guarantee using the poetry virtualenv (if available). 

Note that you can also [configure poe to use a a specific virtualenv or to have no opinion about it](https://poethepoet.natn.io/global_options.html#change-the-executor-type).

**Comment by fritz-trawa** [comment, marked as answer]
Hi just a heads up to thank you for responding, I'll get back to it after my holidays

**Comment by fritz-trawa** [comment]
Reopening as the issue is occurring again

I have added the task

```toml
[tool.poe.tasks.wtf]
shell = "poetry env info -p"
```
and


```toml
[tool.poe.executor]
type = "poetry"
```
And with them

```bash
❯ poetry env info -p
/Users/fritz/Library/Caches/pypoetry/virtualenvs/timeseries-hub-scripts-BBfc8S4B-py3.10

❯ poe wtf
/Users/fritz/Library/Caches/pypoetry/virtualenvs/timeseries-hub-Zz_YmjRr-py3.12
```

3.12 is an old venv, poe for some reason seems to be fixated on using it. Does it cache its settings somewhere?

**Reply by nat-n** [reply]
well that's curious...

Setting `tool.poe.executor.type = "poetry"` should normally have no effect, since that is effectively the default if your project uses poetry.

[The method](https://github.com/nat-n/poethepoet/blob/main/poethepoet/executor/poetry.py#L68) gets called every time that poe needs to know where to find the virtualenv from poetry. As you can see it does use an in memory cache, but it's not persisted anywhere outside the process, so it's just just there to save repeated (and expensive) calls to poetry when running multiple tasks (such as a sequence of tasks).

So somehow poetry must be responding differently when called directly from your shell, vs when called indirectly by poe. The only reason I can think of that this might happen is if you have python 3.10 active in your shell (via pyenv or something?), but if poe is installed via pipx or homebrew in a 3.12 based environment, or perhaps it simply isn't propagating the shell (or directory?) level python version config to the poetry subprocess then poetry defaults to using the 3.12 based venv.

Do you think I'm on the right track here? Could you share some more details about your local environment to help validate this theory? There might be a way to fix it.

**Reply by fritz-trawa** [reply]
We are reaching the limits of my expertise 😄 ... yes I did originally install python with pyenv (and not with `poetry env use python3.10`) then run `poetry shell`. And python 3.12 is my global python (again, set with pyenv). So what happened is 

```bash
❯ python -V
3.12

❯ poetry shell
... do poetry things

# oh shit I forgot to switch python version

❯ poetry env list
# lots of fiddling with env, poetry env remove

❯ pyenv local 3.10
❯ poetry shell

```

I have now manually removed the 3.12 virtual env with `rm -rf` and poe switched to using 3.10. So all it's well now.
I guess this is an edge case that is not worth wasting your time on, I shared in case it provides useful insights. 
Thanks for your help
